torch/csrc/autograd/variable.cpp

```diff
@@ -121,6 +122,20 @@ std::shared_ptr<Function>& VariableViewImpl::get_grad_fn() {
   return _grad_fn;
 }
 
+void VariableViewImpl::set_grad_fn(std::shared_ptr<Function> grad_fn) {
+  if (this->_grad_fn && grad_fn) {
+    // If we have a previous grad_fn then this is an in-place modification
```

Comment:
Views don't have grad functions if they don't require grad, right? So you can have in-place modifications even though you didn't have grad_fn, and the fact that you don't have it doesn't imply that your base still doesn't require grad (it might have been modified in the meantime and you didn't check)

Reply:
If a view doesn't have a grad_fn, then it's a leaf and we don't currently support in-place operations on leaves.

The logic here worries me too, though. Perhaps we should explicitly have VariableViewImpl::rebase(grad_fn),  which is only called for in-place modifications, instead of trying to determine if it's a modification or new grad_fn internally.

Reply:
We do support in-place on leaves if the don't require grad. So how is it finally, is it true that views always have grad_fns or not (my guess is no)?

Reply:
Yeah, this situations that turn on requires_grad in base seem really dangerous. Any operations on the views will see that they don't require grad, so they won't emit gradient for them, but they should reach the base!